pkg/handlers/v1/storage/db.go

```diff
@@ -0,0 +1,309 @@
+			isJoin = true
+		}
+		err := db.insertIPAddress(ctx, val, tx)
+		if err != nil {
+			return err
+		}
+		err = db.insertNetworkChangeEvent(ctx, timestamp, false, isJoin, resourceID, val, nil, tx)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, val := range changes.PublicIPAddresses {
+		isJoin := false
+		if changes.ChangeType == "ADDED" {
+			isJoin = true
+		}
+		err := db.insertIPAddress(ctx, val, tx)
+		if err != nil {
+			return err
+		}
+		// yeah, a nested for loop to establish the relationship of every public IP to every hostname
+		for _, hostname := range changes.Hostnames {
+			err := db.insertNetworkChangeEvent(ctx, timestamp, true, isJoin, resourceID, val, &hostname, tx)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (db *DB) insertIPAddress(ctx context.Context, ipAddress string, tx *sql.Tx) error {
+	// this lib won't give back the last INSERTed row ID, so we don't bother with `RETURNING ...`
+	// See https://stackoverflow.com/questions/34708509/how-to-use-returning-with-on-conflict-in-postgresql
+	if _, err := tx.ExecContext(ctx, fmt.Sprintf(`INSERT INTO %s VALUES ($1) ON CONFLICT DO NOTHING`, tableAWSIPS), ipAddress); err != nil { // nolint
+		return err
+	}
+
+	return nil
+}
+
+func (db *DB) insertHostname(ctx context.Context, hostname string, tx *sql.Tx) error {
+	// this lib won't give back the last INSERTed row ID, so we don't bother with `RETURNING ...`
+	// See https://stackoverflow.com/questions/34708509/how-to-use-returning-with-on-conflict-in-postgresql
+	if _, err := tx.ExecContext(ctx, fmt.Sprintf(`INSERT INTO %s VALUES ($1) ON CONFLICT DO NOTHING`, tableAWSHostnames), hostname); err != nil { // nolint
+		return err
+	}
+
+	return nil
+}
+
+func (db *DB) insertNetworkChangeEvent(ctx context.Context, timestamp time.Time, isPublic bool, isJoin bool, resourceID string, ipAddress string, hostname *string, tx *sql.Tx) error {
+	// Postgres does not auto-create partition tables, so we do it.  We're using 3-month intervals (quarters)
+	monthInterval := 3
+	year := timestamp.Year()
+	fromMonth := (int((timestamp.Month())-1)/monthInterval)*monthInterval + 1 // get the interval of the year we're in, then use the first month of that quarter
+	fromDay := 1
+	toMonth := (int((timestamp.Month())-1)/monthInterval)*monthInterval + monthInterval        // get the interval of the year we're in, then use the last month of that interval
+	toDay := time.Date(year, time.Month(toMonth+1), 0, 0, 0, 0, 0, timestamp.Location()).Day() // time.Date will normalize that toMonth + 1 and 0 day shenanigans
+	from := fmt.Sprintf(`%d-%02d-%02d`, year, fromMonth, fromDay)
+	to := fmt.Sprintf(`%d-%02d-%02d`, year, toMonth, toDay)
+	partitionTableNameSuffix := fmt.Sprintf(`%d_%02dto%02d`, year, fromMonth, toMonth)
+	tableName := fmt.Sprintf(`%s_%s`, tableAWSEventsIPSHostnames, partitionTableNameSuffix)
+
+	if _, err := tx.ExecContext(ctx, fmt.Sprintf(`CREATE TABLE IF NOT EXISTS %s PARTITION OF %s FOR `+  // nolint
```

Comment:
I don't have any data around this but my naive assumption would be that this is an expensive check to run on _every_ insert. Additionally, when this actually results in table creation then my naive expectation is that this is likely an order of magnitude or more slower than an insert especially when factoring in the index creation. Would it be possible to automate this on a schedule instead?